### PR TITLE
prov/efa: refactor RNR support

### PR DESCRIPTION
**Pull request by wzamazon**
This PR contains patches that refactors RNR support

### CODE REVIEW

File: prov/efa/src/rxr/rxr_cq.c [hunk 1]
```diff
@@ -268,18 +268,18 @@ static inline void rxr_cq_queue_pkt(struct rxr_ep *ep,
...
 
-		peer->rnr_timeout_exp = 1;
 		FI_DBG(&rxr_prov, FI_LOG_EP_DATA,
 		       "initializing backoff timeout for peer: %" PRIu64
 		       " timeout: %d rnr_queued_pkts: %d\n",
 		       pkt_entry->addr, peer->timeout_interval,
 		       peer->rnr_queued_pkt_cnt);
 	} else {
-		peer->rnr_timeout_exp++;
+		peer->timeout_interval = MIN(rxr_env.max_timeout,
```
Comment (bwbarrett) [inline prov/efa/src/rxr/rxr_cq.c]: This is only safe if rxr_env.max_timeout is less than 1/2 of MAX_INT, which is a condition we don't enforce anywhere.

File: prov/efa/src/rxr/rxr_cq.c [hunk 2]
```diff
@@ -227,39 +227,93 @@ int rxr_cq_handle_tx_error(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
 	return ofi_cq_write_error(util_cq, &err_entry);
 }
 
-/*
- * Queue a packet on the appropriate list when an RNR error is received.
+/* @brief Queue a packet that encountered RNR error and setup RNR backoff
+ *
+ * We uses an exponential backoff strategy to handle RNR errors.
+ *
+ * `Backoff` means if a peer encountered RNR, an endpoint will
```
Comment (bwbarrett) [inline prov/efa/src/rxr/rxr_cq.c:234]: While maybe defining exponential is overkill, thank you for the detailed comments in this patch.

**Review by bwbarrett — APPROVED** [review]
Patch looks good, although we should really put enforcement of the max interval in initialization.

File: prov/efa/src/rxr/rxr_cq.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c]: I can make the type of `rxr_env.max_timeout` uint32_t, so it is always positive.

Comment (bwbarrett) [reply prov/efa/src/rxr/rxr_cq.c]: what is (MAX_UINT32 / 2 + 2) * 2?  Just put an upper bound for max_timeout that is well above sanity but well below 1/2 MAX_INT and call it a day.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c]: Ah, I see your point now. So you are worried timeout was set to `rxr_env.max_timeout` then overflow when multiplied with 2.

Sure, will add a check in initialization.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c]: Added the check in new revision.

**Pull request merged**